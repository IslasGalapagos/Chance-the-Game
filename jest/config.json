{
  "rootDir": "../",
  "moduleFileExtensions": ["ts", "tsx", "js", "jsx", "json", "node"],
  "setupFiles": ["<rootDir>/jest/setup.js"],
  "testRegex": "(/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$",
  "transform": {
    "^.+\\.tsx?$": "ts-jest"
  }
}
