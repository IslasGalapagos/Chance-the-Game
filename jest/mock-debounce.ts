/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

type TDebouncedF = () => {};

let debouncedF: TDebouncedF;

export const debouncedMock = jest.fn(() => debouncedF());

jest.mock('lodash-es/debounce', () =>
  jest.fn((fn: TDebouncedF) => {
    debouncedF = fn;
    return debouncedMock;
  })
);
