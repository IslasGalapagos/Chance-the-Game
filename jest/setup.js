import {configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import jsdom from 'jsdom';

configure({adapter: new Adapter()});

const {JSDOM} = jsdom;
const dom = new JSDOM(
  '<!DOCTYPE html><html><body><div id="root"></div></body></html>'
);

window.eval(`
  Object.defineProperty(window.document.body, 'clientHeight', {
    get() {
      return window.innerHeight;
    }
  });
`);

global.window = window;
global.document = window.document;

global.window.resizeTo = (width, height) => {
  global.window.innerWidth = width || global.window.innerWidth;
  global.window.innerHeight = width || global.window.innerHeight;
  global.window.dispatchEvent(new Event('resize'));
};
