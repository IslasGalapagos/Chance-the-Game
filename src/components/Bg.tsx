import * as React from 'react';
import debounce from 'lodash-es/debounce';
import styled from 'react-emotion';

const StyledCanvas = styled('canvas')`
  position: absolute;
  z-index: 0;
`;

interface GridConfig {
  readonly ROWS: number;
  readonly COLS: number;
  readonly COLOR: string;
}

interface State {
  readonly width: number;
  readonly height: number;
}

class Bg extends React.PureComponent<{}, State> {
  readonly state: State;

  readonly gridConfig: GridConfig = {
    ROWS: 6,
    COLS: 3,
    COLOR: '#F2FEFD'
  };

  onResize = debounce(() => {
    this.setState(this.getSizes());

    this.draw();
  }, 200);

  private canvasRef: React.RefObject<HTMLCanvasElement>;

  constructor(props: {}) {
    super(props);

    this.canvasRef = React.createRef();

    this.state = this.getSizes();
  }

  componentDidMount() {
    window.addEventListener('resize', this.onResize);

    this.draw();
  }

  getSizes() {
    return {
      width: window.innerWidth,
      height: document.body.clientHeight
    };
  }

  draw() {
    // TODO: write visual test ?
    // TODO: create more complicated thing â€“ interactive or animated
    const canvas = this.canvasRef.current;

    if (canvas === null) {
      return;
    }

    const ctx = canvas.getContext('2d');

    if (ctx === null) {
      return;
    }

    const {ROWS, COLS, COLOR} = this.gridConfig;
    const {width, height} = this.state;

    const rectWidth = Math.floor(width / COLS);
    const lastRectWidth = width - rectWidth * (COLS - 1);
    const rectHeight = Math.floor(height / ROWS);
    const lastRectHeight = height - rectHeight * (ROWS - 1);

    ctx.fillStyle = COLOR;

    for (let i = 0; i < ROWS; i += 1) {
      for (let j = 0; j < COLS; j += 1) {
        if ((i % 2 === 0 && j % 2 === 0) || (i % 2 !== 0 && j % 2 !== 0)) {
          ctx.fillRect(
            j * rectWidth,
            i * rectHeight,
            j < COLS - 1 ? rectWidth : lastRectWidth,
            i < ROWS - 1 ? rectHeight : lastRectHeight
          );
        }
      }
    }
  }

  render() {
    return (
      <StyledCanvas
        width={`${this.state.width}px`}
        height={`${this.state.height}px`}
        innerRef={this.canvasRef}
      />
    );
  }
}

export default Bg;
