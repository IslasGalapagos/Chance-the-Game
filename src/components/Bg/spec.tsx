/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import {shallow} from 'enzyme';
import * as React from 'react';
import * as testRenderer from 'react-test-renderer';
import debounce from 'lodash-es/debounce';
import {debouncedMock} from '../../../jest/mock-debounce';

import Bg, {StyledCanvas} from '.';

describe('<Bg/>', () => {
  beforeEach(() => {
    (debounce as any).mockClear();
    debouncedMock.mockClear();
  });

  it('renders correctly', () => {
    const treeAndStyles = testRenderer.create(<Bg />).toJSON();
    expect(treeAndStyles).toMatchSnapshot();
  });

  // REFS
  it('has a refs', () => {
    const component = shallow(<Bg />);
    expect(component.instance()).toHaveProperty('canvasRef');
    expect(component.instance()).toHaveProperty('logoRef');
  });

  // CONFIG
  it('has config with data for drawing', () => {
    const component = shallow(<Bg />).instance() as Bg;
    expect(component.gridConfig).toEqual(
      expect.objectContaining({
        ROWS: expect.any(Number),
        COLS: expect.any(Number),
        COLOR: expect.any(String)
      })
    );
  });

  // STATE
  it('has initial state with window sizes & renders with these sizes', () => {
    const component = shallow(<Bg />);
    expect(component.state('width')).toEqual(window.innerWidth);
    expect(component.state('height')).toEqual(
      window.document.body.clientHeight
    );
    const canvas = component.find(StyledCanvas).render();
    expect(canvas.prop('width')).toEqual(`${window.innerWidth}px`);
    expect(canvas.prop('height')).toEqual(
      `${window.document.body.clientHeight}px`
    );
  });

  // EVENT HANDLERS
  it('changes state & rerenders with new sizes on window.resize', () => {
    const component = shallow(<Bg />);
    let canvas = component.find(StyledCanvas).render();
    const sizeForTest = 100;

    expect(component.state('width')).not.toEqual(sizeForTest);
    expect(component.state('height')).not.toEqual(sizeForTest);
    expect(canvas.prop('width')).not.toEqual(`${sizeForTest}px`);
    expect(canvas.prop('height')).not.toEqual(`${sizeForTest}px`);

    window.resizeTo(sizeForTest, sizeForTest);

    canvas = component.find(StyledCanvas).render();
    expect(component.state('width')).toEqual(sizeForTest);
    expect(component.state('height')).toEqual(sizeForTest);
    expect(canvas.prop('width')).toEqual(`${sizeForTest}px`);
    expect(canvas.prop('height')).toEqual(`${sizeForTest}px`);
  });

  it('has state.logoIsLoad, that changes after logo image is loaded', () => {
    const spyOnLogoLoad = jest.spyOn(Bg.prototype, 'onLogoLoad');
    const component = shallow(<Bg />);
    expect(component.state('logoIsLoad')).toEqual(false);
    expect(spyOnLogoLoad).not.toBeCalled();
    component.find('img').simulate('load');
    expect(spyOnLogoLoad).toBeCalledTimes(1);
    expect(component.state('logoIsLoad')).toEqual(true);
  });

  // METHODS & LIFECYCLES
  it('calls draw when logo loaded', () => {
    const spyDraw = jest.spyOn(Bg.prototype, 'draw');
    const component = shallow(<Bg />);
    spyDraw.mockClear();
    (component.instance() as Bg).onLogoLoad();
    expect(spyDraw).toHaveBeenCalled();
  });

  it('calls draw on mount', () => {
    const spyCDM = jest.spyOn(Bg.prototype, 'componentDidMount');
    const spyDraw = jest.spyOn(Bg.prototype, 'draw');
    shallow(<Bg />);
    expect(spyCDM).toHaveBeenCalled();
    expect(spyDraw).toHaveBeenCalled();
  });

  it('calls draw in onResize with debounce', () => {
    const spyDraw = jest.spyOn(Bg.prototype, 'draw');
    expect(debounce).not.toHaveBeenCalled();
    const component = shallow(<Bg />).instance() as Bg;
    expect(debounce).toHaveBeenCalledTimes(1);
    expect(debouncedMock).not.toHaveBeenCalled();
    spyDraw.mockClear();
    component.onResize();
    expect(debouncedMock).toHaveBeenCalledTimes(1);
    expect(spyDraw).toHaveBeenCalledTimes(1);
  });

  it('calls onResize on window.resize', () => {
    const component = shallow(<Bg />).instance() as Bg;
    const spyOnResize = jest.spyOn(component, 'onResize');
    expect(spyOnResize).not.toHaveBeenCalled();
    window.resizeTo(100, 100);
    expect(spyOnResize).toHaveBeenCalledTimes(1);
  });

  it('removes resize handler on unmount', () => {
    window.removeEventListener = jest.fn();
    const component = shallow(<Bg />);
    expect(window.removeEventListener).not.toHaveBeenCalled();
    component.unmount();
    expect(window.removeEventListener).toHaveBeenCalled();
  });
});
