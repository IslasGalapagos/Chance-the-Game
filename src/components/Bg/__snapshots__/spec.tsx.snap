// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Bg/> renders correctly 1`] = `
Array [
  <img
    hidden={true}
    onLoad={[Function]}
    src="./images/chance.svg"
  />,
  .emotion-0 {
  position: absolute;
  z-index: 0;
}

<canvas
    className="emotion-0"
    height="768px"
    width="1024px"
  />,
]
`;
